File: sort/rotatedarrsearch/RotatedArraySearch.py
class RotatedArraySearch:
    def search_in_rotated_array(self, nums, target):
        left_index, right_index = 0, len(nums) - 1

        while left_index <= right_index:
            mid_index = left_index + (right_index - left_index) // 2

            # Check if the target is found at mid_index
            if nums[mid_index] == target:
                return mid_index

            # Check if the left half is sorted
            if nums[left_index] <= nums[mid_index]:
                # Target is in the sorted left half
                if nums[left_index] <= target < nums[mid_index]:
                    right_index = mid_index - 1  # Search in left half
                else:
                    left_index = mid_index + 1  # Search in right half
            # The right half must be sorted
            else:
                # Target is in the sorted right half
                if nums[mid_index] < target <= nums[right_index]:
                    left_index = mid_index + 1  # Search in right half
                else:
                    right_index = mid_index - 1  # Search in left half

        return -1  # Target not found

# Example usage
if __name__ == "__main__":
    searcher = RotatedArraySearch()
    print(searcher.search_in_rotated_array([4, 5, 6, 7, 0, 1, 2], 0))  # Output: 4
    print(searcher.search_in_rotated_array([4, 5, 6, 7, 0, 1, 2], 3))  # Output: -1

File: sort/rotatedarrsearch/RotatedArraySearch.java
package rotatedarrsearch;

class RotatedArraySearch {
    public int searchInRotatedArray(int[] nums, int target) {
        int leftIndex = 0, rightIndex = nums.length - 1;

        while (leftIndex <= rightIndex) {
            int midIndex = leftIndex + (rightIndex - leftIndex) / 2;

            // Check if the target is found at midIndex
            if (nums[midIndex] == target) {
                return midIndex;
            }

            // Check if the left half is sorted
            if (nums[leftIndex] <= nums[midIndex]) {
                // Target is in the sorted left half
                if (nums[leftIndex] <= target && target < nums[midIndex]) {
                    rightIndex = midIndex - 1; // Search in left half
                } else {
                    leftIndex = midIndex + 1; // Search in right half
                }
            }
            // The right half must be sorted
            else {
                // Target is in the sorted right half
                if (nums[midIndex] < target && target <= nums[rightIndex]) {
                    leftIndex = midIndex + 1; // Search in right half
                } else {
                    rightIndex = midIndex - 1; // Search in left half
                }
            }
        }

        return -1; // Target not found
    }

    public static void main(String[] args) {
        RotatedArraySearch searcher = new RotatedArraySearch();
        System.out.println(searcher.searchInRotatedArray(new int[]{4, 5, 6, 7, 0, 1, 2}, 0));  // Output: 4
        System.out.println(searcher.searchInRotatedArray(new int[]{4, 5, 6, 7, 0, 1, 2}, 3));  // Output: -1
    }
}
